joint_limits_interface/include/joint_limits_interface/joint_limits_interface.hpp

```diff
@@ -0,0 +1,665 @@
+      vel_high = std::min(prev_cmd_ + limits_.max_acceleration * dt, limits_.max_velocity);
+    } else {
+      vel_low = -limits_.max_velocity;
+      vel_high = limits_.max_velocity;
+    }
+
+    // Saturate velocity command according to limits
+    const double vel_cmd = rcppmath::clamp(
+      jcmdh_.get_cmd(),
+      vel_low,
+      vel_high);
+    jcmdh_.set_cmd(vel_cmd);
+
+    // Cache variables
+    prev_cmd_ = jcmdh_.get_cmd();
+  }
+
+private:
+  hardware_interface::JointStateHandle jh_;
+  hardware_interface::JointCommandHandle jcmdh_;
+  joint_limits_interface::JointLimits limits_;
+
+  double prev_cmd_;
+};
+
+/** \brief A handle used to enforce position, velocity, and acceleration limits of a
+  * velocity-controlled joint.
+  */
+class VelocityJointSoftLimitsHandle
+{
+public:
+  VelocityJointSoftLimitsHandle() {}
+  VelocityJointSoftLimitsHandle(
+    const hardware_interface::JointStateHandle & jh,
+    const hardware_interface::JointCommandHandle & jcmdh,
+    const joint_limits_interface::JointLimits & limits,
+    const joint_limits_interface::SoftJointLimits & soft_limits)
+  {
+    jh_ = jh;
+    jcmdh_ = jcmdh;
+    limits_ = limits;
+    soft_limits_ = soft_limits;
+    if (limits.has_velocity_limits) {
+      max_vel_limit_ = limits.max_velocity;
+    } else {
+      max_vel_limit_ = std::numeric_limits<double>::max();
+    }
+  }
+
+  /** \return Joint name. */
+  std::string get_name() const {return jh_.get_name();}
+
+  /**
+   * \brief Enforce position, velocity, and acceleration limits for a velocity-controlled joint
+   * subject to soft limits.
+   *
+   * \param period Control period.
+   */
+  void enforceLimits(const rclcpp::Duration & period)
+  {
+    double min_vel, max_vel;
+    if (limits_.has_position_limits) {
+      // Velocity bounds depend on the velocity limit and the proximity to the position limit.
+      const double pos = jh_.get_position();
+      min_vel = rcppmath::clamp(
+        -soft_limits_.k_position * (pos - soft_limits_.min_position),
+        -max_vel_limit_, max_vel_limit_);
+      max_vel = rcppmath::clamp(
+        -soft_limits_.k_position * (pos - soft_limits_.max_position),
+        -max_vel_limit_, max_vel_limit_);
+    } else {
+      min_vel = -max_vel_limit_;
+      max_vel = max_vel_limit_;
+    }
+
+    if (limits_.has_acceleration_limits) {
+      const double vel = jh_.get_velocity();
+      const double delta_t = period.seconds();
+      min_vel = std::max(vel - limits_.max_acceleration * delta_t, min_vel);
+      max_vel = std::min(vel + limits_.max_acceleration * delta_t, max_vel);
+    }
+
+    jcmdh_.set_cmd(rcppmath::clamp(jcmdh_.get_cmd(), min_vel, max_vel));
+  }
+
+private:
+  hardware_interface::JointStateHandle jh_;
+  hardware_interface::JointCommandHandle jcmdh_;
+  joint_limits_interface::JointLimits limits_;
+  joint_limits_interface::SoftJointLimits soft_limits_;
+  double max_vel_limit_;
+};
+
+#if 0
```

Comment:
It seems this code is never used. Do we still need it?

Reply:
added a TODO